packages/firestore/test/unit/specs/bundle_spec.test.ts

```diff
@@ -0,0 +1,357 @@
+      key: docA.key,
+      readTime: 3000,
+      createTime: 1999,
+      updateTime: 2999,
+      content: { key: { stringValue: 'b' } }
+    });
+
+    return spec()
+      .userListens(query1)
+      .watchAcksFull(query1, 1000, docA)
+      .expectEvents(query1, { added: [docA] })
+      .loadBundle(bundleString)
+      .expectEvents(query1, { modified: [docAChanged] });
+  });
+
+  specTest('Newer deleted docs from bundles should delete cache.', [], () => {
+    const query1 = Query.atPath(path('collection'));
+    const docA = doc('collection/a', 1000, { key: 'a' });
+
+    const bundleString = bundleWithDocument({
+      key: docA.key,
+      readTime: 3000
+    });
+
+    return spec()
+      .userListens(query1)
+      .watchAcksFull(query1, 1000, docA)
+      .expectEvents(query1, { added: [docA] })
+      .loadBundle(bundleString)
+      .expectEvents(query1, { removed: [docA] });
+  });
+
+  specTest('Older deleted docs from bundles should do nothing.', [], () => {
+    const query1 = Query.atPath(path('collection'));
+    const docA = doc('collection/a', 1000, { key: 'a' });
+
+    const bundleString = bundleWithDocument({
+      key: docA.key,
+      readTime: 999
+    });
+
+    return (
+      spec()
+        .userListens(query1)
+        .watchAcksFull(query1, 1000, docA)
+        .expectEvents(query1, { added: [docA] })
+        // No events are expected here.
+        .loadBundle(bundleString)
+    );
+  });
+
+  specTest(
+    'Newer docs from bundles should raise snapshot only when watch catches up with acknowledged writes.',
+    [],
+    () => {
+      const query = Query.atPath(path('collection'));
+      const docA = doc('collection/a', 250, { key: 'a' });
+
+      const bundleString1 = bundleWithDocument({
+        key: docA.key,
+        readTime: 500,
+        createTime: 250,
+        updateTime: 500,
+        content: { key: { stringValue: 'b' } }
+      });
+
+      const bundleString2 = bundleWithDocument({
+        key: docA.key,
+        readTime: 1001,
+        createTime: 250,
+        updateTime: 1001,
+        content: { key: { stringValue: 'fromBundle' } }
+      });
+      return (
+        spec()
+          .withGCEnabled(false)
+          .userListens(query)
+          .watchAcksFull(query, 250, docA)
+          .expectEvents(query, {
+            added: [doc('collection/a', 250, { key: 'a' })]
+          })
+          .userPatches('collection/a', { key: 'patched' })
+          .expectEvents(query, {
+            modified: [
+              doc(
+                'collection/a',
+                250,
+                { key: 'patched' },
+                { hasLocalMutations: true }
+              )
+            ],
+            hasPendingWrites: true
+          })
+          .writeAcks('collection/a', 1000)
+          // loading bundleString1 will not raise snapshots, because it is before
+          // the acknowledged mutation.
+          .loadBundle(bundleString1)
```

Comment:
Nit: Please use descriptive name for `bundleString1` and `bundleString2`. You might even be able to get rid of this comment.